[#7454] Add basic copy to public chat empty screen state

bitsikka opened the pull request:
Fixes #7454 

### Summary:

tested on iOS Simulator and Android AVD

### Review notes:
Guessed the design spec - need input

status: ready

bitsikka committed:
[#7454] fix add basic copy to public chat empty screen state + related issues/feature regarding chat message-views intro screens for all chats types

[#7454] fix add basic copy to public chat empty screen state + chat messages-views intro screens for all chats

Signed-off-by: Igor Mandrigin <i@mandrigin.ru>

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status.html))?

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 7a99191f | [#1](https://ci.status.im/job/status-react/job/prs/job/macos/job/PR-7663/1/) | 2019-03-04 19:12:19 | ~14 min | `macos` | [:package: dmg](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190304-185752-7a9919-pr.dmg) |

yenda commented on src/status_im/ui/screens/chat/views.cljs:
```diff
@@ -109,9 +110,48 @@
+        message
+        (i18n/label :t/share-public-chat-text {:link link})
+
+        public-chat-name-style
```
if you don't inline the style directly it's convention to put it in styles.cljs next to the views

bitsikka replied:
got it!

bitsikka replied:
done!

yenda commented on src/status_im/ui/screens/chat/views.cljs:
```diff
@@ -109,9 +110,48 @@
-   [react/text {:style style/empty-chat-text}
-    (i18n/label :t/empty-chat-description)]])
+  (let [{:keys [color chat-id chat-name]}
+        @(re-frame/subscribe [:chats/current-chat])
```
if possible, try to pass parameters to the function rather than subscribe within

bitsikka replied:
I see
gonna subscribe it in letsubs in message-view component and pass it to empty-chat-container

bitsikka replied:
done!

yenda commented on src/status_im/ui/screens/chat/views.cljs:
```diff
@@ -109,9 +110,48 @@
+        @(re-frame/subscribe [:chats/current-chat])
+
+        link
+        (universal-links/generate-link :public-chat :external chat-id)
```
computations should be done in subscriptions

bitsikka replied:
I see. I understand why computations should be done in subscriptions but not sure about which subscription. Like in `:chats/current-chat`? or new subscription?

Can you please elaborate?

bitsikka replied:
it could be calculated and stored in app-db at the time of chat creation right? But I kinda followed what was done in chat actions which also has share chat item.

https://github.com/status-im/status-react/blob/aa0c45738200d64b39f95955b2e518740e3080e1/src/status_im/ui/screens/chat/actions.cljs#L15

I could move the calculation to private function like it does.

bitsikka replied:
did it as in chat actions
done!

yenda replied:
Ideally in :chats/current-chat subscription if there is no messages (chat empty), you add a compute and add the link key, no need to persist it to app-db. Chat actions code isn't a good example it should be refactored at some point

bitsikka replied:
ok got it!
another update coming

bitsikka replied:
done!

hesterbruikman commented:
@bitsikka checking the builds now

hesterbruikman commented:
## iOS Before & After
![before_after ios](https://user-images.githubusercontent.com/35233054/53795528-7841be80-3f32-11e9-838a-6d27b16ad1df.png)

Trying to get the same for Android, build keeps crashing on my phone:( Anyoneelse with Android that can post a before/after?

errorists commented on src/status_im/ui/screens/chat/views.cljs:
```diff
@@ -109,9 +110,48 @@
+                 :font-weight   :bold
+                 :line-height   28
+                 :margin-bottom 10
+                 :color         colors/black}}
```
Corrections incoming, font-size should be a 22, bottom margin a 8.

bitsikka replied:
done!

errorists commented on src/status_im/ui/screens/chat/views.cljs:
```diff
@@ -109,9 +110,48 @@
+
+        public-chat-icon-style
+        {:margin-bottom    25
+         :width            130
```
width and height should be 120, bottom margin 24

bitsikka replied:
done!

errorists commented on src/status_im/ui/screens/chat/views.cljs:
```diff
@@ -109,9 +110,48 @@
+
+        public-chat-icon-text-style
+        {:style {:color       colors/white
+                 :font-size   56
```
font-size should be 52

bitsikka replied:
done!

bitsikka commented:
I had not accounted for the fact that private group chat also used the same `empty-chat-container` and was problematic when sharing the chat. Fixed in the latest update to the PR along with changes requested.

yenda commented:
@bitsikka thanks looks pretty nice, you implemented a feature and improved the codebase at the same time :)

bitsikka commented:
> @bitsikka thanks looks pretty nice, you implemented a feature and improved the codebase at the same time :)

my pleasure! 😄 

in the last force-push I tried to follow your instruction to the tee - don't know if that was necessary. I guess it allows more flexibility if needed to be used elsewhere.

Thanks for all the guidance! I learned a lot!

flexsurfer commented on src/status_im/ui/screens/chat/styles/main.cljs:
```diff
@@ -238,6 +238,29 @@
    :color         colors/black})
 
+(def empty-chat-name
+  {:style {:font-size     22
```
could you please move style key outside?

bitsikka replied:
done!

flexsurfer commented on src/status_im/ui/screens/chat/styles/main.cljs:
```diff
@@ -238,6 +238,29 @@
+   :background-color color})
+
+(def empty-chat-icon-text
+  {:style {:color       colors/white
```
could you please move style key outside?

bitsikka replied:
done!

flexsurfer commented on src/status_im/ui/screens/chat/views.cljs:
```diff
@@ -107,6 +107,22 @@
+  [react/view style/empty-chat-container
+   [react/view (style/empty-chat-icon 120 color)
+    [react/text style/empty-chat-icon-text
+     (clojure.string/capitalize (first chat-id))]]
```
could you use require :as string here?

bitsikka replied:
done!

errorists commented:
@bitsikka thanks for working on this! @yenda can we create a separate issue to get all those different chat headers a part of chat content and not a separate view? Here's a [Figma of what I mean ](https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=2%3A341) The intent is these are like a book cover, they should stay there on top and scroll with the rest of messages.

bitsikka commented:
> @bitsikka thanks for working on this! @yenda can we create a separate issue to get all those different chat headers a part of chat content and not a separate view? Here's a [Figma of what I mean ](https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=2%3A341) The intent is these are like a book cover, they should stay there on top and scroll with the rest of messages.

cool designs! 👍 I see what you mean.

yenda commented:
@errorists it can be added as a first element to the message list

errorists commented:
@yenda cool! maybe @siphiuel can add it as part of the TtT chat PR, this header is heavily featured there?

bitsikka commented:
> @yenda cool! maybe @siphiuel can add it as part of the TtT chat PR, this header is heavily featured there?

I take it that, what @hesterbruikman said in #7454 in the description of the issue:
> We can offer a quick solution while we wait for the epics 'Onboarding' and 'Empty state screens' to progress.

the wait is over

If so, and by the look of Figma design linked, wait is indeed over, and if everybody is busy, then I could work on it. I understand it will need some restructuring of UI and care is needed to touch only what is needed, and to not break anything else.

bitsikka commented:
> @errorists it can be added as a first element to the message list

I see what you mean

churik commented:
@bitsikka thank you for your contribution!

Some questions/ suggestions:
1) @hesterbruikman WDYT should basic copy to empty public chat view shown whilst messages are fetching?
For now, messages are not fetching in seconds, it can take some time and after they are fetched this empty placeholder will be replaced with chat history.
IMO this placeholder should be displayed only after all history was fetched and there is no messages for last 24h  for sure.

2) @errorists about view when keyboard is opened: IMO this view should be scrollable or have a responsive design to adjust window size because for now, it looks different for IOS and Android.
Don't think it is actually a big deal but anyway would be nice to know your opinion.
On screenshot on the right LG V20 (Android 8), on the left IPhone 7(IOS 11.4.1)
![ios_andr](https://user-images.githubusercontent.com/4557972/53873695-2e73d980-4001-11e9-9739-7971e6980c08.png)

bitsikka commented:
> @bitsikka thank you for your contribution!

my pleasure 😊 

> Some questions/ suggestions:

I can definitely look into these very sensible suggestions together with implementation of all empty screen states 👍

hesterbruikman commented:
>     1. @hesterbruikman WDYT should basic copy to empty public chat view shown whilst messages are fetching?
>        For now, messages are not fetching in seconds, it can take some time and after they are fetched this empty placeholder will be replaced with chat history.
>        IMO this placeholder should be displayed only after all history was fetched and there is no messages for last 24h  for sure.

Agreed, So it sounds like we need an empty loading state screen in addition to the 'No messages in the last 24h' screen above. I will create a another issue to look at an empty loading screen. For this issue, indeed let's only show the above screen once fetching for the last 24h is done. 

@bitsikka Great work, thank you! Can you update so that the screen is only shown after all fetching for the last 24h is done? For the empty screen state/loading I'll create another issue/new bounty. 

We might need a design for that and also lets make sure you get properly rewarded before we ask you to implement 20 new UI designs:P

errorists commented:
@churik thanks for reviewing, you're right it should be scrollable but even more, it should be a part of regular chat history. Like Eric said, let's think of it as the first message on the list. The exact same header will be displayed for 1-1, group and public chats, just having different text in subtitle. If you want to visualise it, just take a peek at the [Chat Figma](https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=3%3A1031), it's all there. 🙏

hesterbruikman commented:
> @bitsikka Great work, thank you! Can you update so that the screen is only shown after all fetching for the last 24h is done? For the empty screen state/loading I'll create another issue/new bounty.

Small correction @bitsikka, as part of this bounty, could you remove the text "There are no messages in this chat yet" from the loading screen. Leaving the screen empty with only the existing loading 'nightrider' at the top below the header? This empty screen would then be replaced once the fetching process is done and there have not been messages in the last 24h.

bitsikka commented:
> > @bitsikka Great work, thank you! Can you update so that the screen is only shown after all fetching for the last 24h is done? For the empty screen state/loading I'll create another issue/new bounty.
> 
> Small correction @bitsikka, as part of this bounty, could you remove the text "There are no messages in this chat yet" from the loading screen. Leaving the screen empty with only the existing loading 'nightrider' at the top below the header? This empty screen would then be replaced once the fetching process is done and there have not been messages in the last 24h.

Can do 👍. Looking into this now.

hesterbruikman commented:
@bitsikka created an issue for the loading screen. Might take some time to get it on Gitcoin given time difference. https://github.com/status-im/status-react/issues/7688

bitsikka commented:
> @bitsikka created an issue for the loading screen. Might take some time to get it on Gitcoin given time difference. #7688

great!
I am still working on empty "without text" loading screen and finding that there is more to this problem than I had initially envisioned. I will get to it though - it will help with the new issue too for me to get the mental model proper/straight.

bitsikka commented:
Update: The last request on this issue and extended PR issue #7688 is so tightly related and deeply inseparable, that it just made sense to decided to include the extended feature of #7688 to be included in update to this PR. That was few days ago.

As I got it done locally(both this and #7688) it just occurred to me that, although already very close, it was still going to be less than what is intended in the Figma designs. That was bothering.

Since it was already so close to delivering the actual complete feature according to the Figma designs, it just made sense to decide to include the complete feature including not just Public chat, but also 1-1, and Group chat.

Hence, I'm going for it.

Just makes sense to just do it.

I should be able to get it done over the weekend.

bitsikka commented:
Note:

 - This latest update will need code review again as it invalidates most of code that came before it
 - It is one big PR - sorry for that, but the nature of the feature required it
 - The implementation for group chat is a bit different from Figma designs
 - Fixes a number of other issues i encountered including #6850 and #7688 and the original #7454

more notes a bit later

churik commented:
@bitsikka nice! re-requested review from developers.

churik commented:
seems that GitHub bot always moved this to `To test` column... @yenda @flexsurfer appreciate if you can take a look.

bitsikka commented:
ping @errorists @hesterbruikman

cammellos commented on src/status_im/mailserver/core.cljs:
```diff
@@ -523,17 +531,33 @@
+      (if (and (= :chat view-id)
+               (:public? (get chats current-chat-id))
+               (= requestID (get-in chats [current-chat-id :join-time-mail-request-id])))
+        (if (or (= lastEnvelopeHash "0x0000000000000000000000000000000000000000000000000000000000000000"))
```
Or is not needed here

hesterbruikman commented:
@bitsikka Thanks for going ahead group chat and existing issues as well!

Can you please share a link to a build I can test (Android and iOS) and/or screenshots?

bitsikka commented:
> @bitsikka Thanks for going ahead group chat and existing issues as well!
> 
> Can you please share a link to a build I can test (Android and iOS) and/or screenshots?

@hesterbruikman although build 11 (just pushed with minor corrections identified earlier) is still being built build 10 should be ok to test. Here they are:

> ### Jenkins Builds
> ✔️	[ae5632e](https://github.com/status-im/status-react/commit/ae5632efeb669b7dec96696abc635a1729f6eef5)	[#10](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7663/10/)	2019-03-18 11:54:43	~14 min	`android-e2e`	[📦 apk](https://status-im.ams3.digitaloceanspaces.com/StatusIm-190318-113948-ae5632-e2e.apk)
> ✔️	[ae5632e](https://github.com/status-im/status-react/commit/ae5632efeb669b7dec96696abc635a1729f6eef5)	[#10](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-7663/10/)	2019-03-18 11:55:38	~15 min	`ios`	[📦 ipa](https://i.diawi.com/JN7iXJ)
> ✔️	[ae5632e](https://github.com/status-im/status-react/commit/ae5632efeb669b7dec96696abc635a1729f6eef5)	[#10](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-7663/10/)	2019-03-18 11:57:20	~17 min	`android`	[📦 apk](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190318-113951-ae5632-pr.apk)

yenda commented:
@bitsikka good job on spotting mailserver/fetching? inconsistencies. After discussing with @cammellos I think the best solution would be to have it rely on a value separated from pending-requests. However this shouldn't be part of this PR. 

@StatusSceptre @hesterbruikman could we make a bounty for this?
`Fix mailserver/fetching? inconsistencies` There should be a 1 second delay between the moment the last request gets completed and the mailserver/fetching? subscription becomes false.

bitsikka commented:
> @bitsikka good job on spotting mailserver/fetching? inconsistencies. After discussing with @cammellos I think the best solution would be to have it rely on a value separated from pending-requests. However this shouldn't be part of this PR.
> 
> @StatusSceptre @hesterbruikman could we make a bounty for this?
> `Fix mailserver/fetching? inconsistencies` There should be a 1 second delay between the moment the last request gets completed and the mailserver/fetching? subscription becomes false.

@yenda thanks!

After the decision I made to go towards the end goal according to figma designs, instead of just temporary workaround as the original issue stated, I realized there had to be few other things(as you have certainly noticed) I would have to wrangle in order to make it work semi-decently.

### Some more notes:

Bugs (other than noted above)
 - there is/was a bug where in particular chat, if you touch on the datemark separator(e.g. "Today", "Yesterday") while beginning to scroll - it wouldn't scroll - fixed
 - there is/was an issue, when a chat is already active/joined and you go back to + -> Join the same chat, it would unnecessarily go through new chat processes - partially fixed (there are other places where unnecessary processing could be avoided)

Extra things needed to be wrangled
 - Content-Layout-Height calculation for FlatList component header where the intro screen elements live now
 - Set-Once Event handler
 - Get-Inviter-Name to be used once user joins group chat

 Design notes:
 - description text that has variables do not have differentiating change in style (color) at the moment
 - group chat intro screens are not exactly according to the designs because of the way group chats work at the moment

bitsikka commented:
Latest Jenkins Builds for convenience

> ✔️	[798c34d](https://github.com/status-im/status-react/commit/798c34df2a840dce4abff073379bbaff1833c23b)	[#11](https://ci.status.im/job/status-react/job/prs/job/macos/job/PR-7663/11/)	2019-03-18 15:19:24	~15 min	`macos`	[📦 dmg](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190318-150417-798c34-pr.dmg)
> ✔️	[798c34d](https://github.com/status-im/status-react/commit/798c34df2a840dce4abff073379bbaff1833c23b)	[#11](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-7663/11/)	2019-03-18 15:20:03	~15 min	`ios`	[📦 ipa](https://i.diawi.com/tAD7Hq)
> ✔️	[798c34d](https://github.com/status-im/status-react/commit/798c34df2a840dce4abff073379bbaff1833c23b)	[#11](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7663/11/)	2019-03-18 15:23:18	~19 min	`android-e2e`	[📦 apk](https://status-im.ams3.digitaloceanspaces.com/StatusIm-190318-150348-798c34-e2e.apk)
> ✔️	[798c34d](https://github.com/status-im/status-react/commit/798c34df2a840dce4abff073379bbaff1833c23b)	[#11](https://ci.status.im/job/status-react/job/prs/job/windows/job/PR-7663/11/)	2019-03-18 15:25:10	~20 min	`windows`	[📦 exe](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190318-150422-798c34-pr.exe)
> ✔️	[798c34d](https://github.com/status-im/status-react/commit/798c34df2a840dce4abff073379bbaff1833c23b)	[#11](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-7663/11/)	2019-03-18 15:28:47	~24 min	`android`	[📦 apk](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190318-150356-798c34-pr.apk)
> ✔️	[798c34d](https://github.com/status-im/status-react/commit/798c34df2a840dce4abff073379bbaff1833c23b)	[#11](https://ci.status.im/job/status-react/job/prs/job/linux/job/PR-7663/11/)	2019-03-18 15:32:59	~28 min	`linux`	[📦 App](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190318-150413-798c34-pr.AppImage)

yenda commented:
> there is/was a bug where in particular chat, if you touch on the datemark separator(e.g. "Today", "Yesterday") while beginning to scroll - it wouldn't scroll - fixed
there is/was an issue, when a chat is already active/joined and you go back to + -> Join the same chat, it would unnecessarily go through new chat processes - partially fixed (there are other places where unnecessary processing could be avoided)

could you extract these 2 as well in separate PR, they will be easier to merge and this one will be easier to review. Also we try to have commits that focus on solving one thing at a time because it makes git bisects easier when looking for the origin of a bug

hesterbruikman commented:
thanks @bitsikka I managed to see the screens on iOS iPhone 8 in public chat. Great improvement to old situation.

Some minor details on the visual/layout, that I hope are not major to fix. Also would like to ask @errorists for clearer specs or a thumbs up as I trust his eyes for visual design more than my own:)

The content (circle and copy) looks slightly off-centre. By about 5px to the left on the x-axis. (See black cross on the image below for reference). The circle itself is bigger on the loading screen than the empty screen state, original is 120x120 for both. https://www.figma.com/file/PK289lt1UWIC7zZNVqFFlB09/Untitled?node-id=0%3A1

![Chat empty states](https://user-images.githubusercontent.com/35233054/54549380-3c830c00-49aa-11e9-8a32-b493641938bf.png)

bitsikka commented:
> > there is/was a bug where in particular chat, if you touch on the datemark separator(e.g. "Today", "Yesterday") while beginning to scroll - it wouldn't scroll - fixed
> > there is/was an issue, when a chat is already active/joined and you go back to + -> Join the same chat, it would unnecessarily go through new chat processes - partially fixed (there are other places where unnecessary processing could be avoided)
> 
> could you extract these 2 as well in separate PR, they will be easier to merge and this one will be easier to review. Also we try to have commits that focus on solving one thing at a time because it makes git bisects easier when looking for the origin of a bug

Got it - will extract them soon

bitsikka commented:
> thanks @bitsikka I managed to see the screens on iOS iPhone 8 in public chat. Great improvement to old situation.
> 
> Some minor details on the visual/layout, that I hope are not major to fix. Also would like to ask @errorists for clearer specs or a thumbs up as I trust his eyes for visual design more than my own:)
> 
> The content (circle and copy) looks slightly off-centre. By about 5px to the left on the x-axis. (See black cross on the image below for reference). The circle itself is bigger on the loading screen than the empty screen state, original is 120x120 for both. https://www.figma.com/file/PK289lt1UWIC7zZNVqFFlB09/Untitled?node-id=0%3A1
> 
> ![Chat empty states](https://user-images.githubusercontent.com/35233054/54549380-3c830c00-49aa-11e9-8a32-b493641938bf.png)

will see about these soon

churik commented:
@bitsikka  so waiting for approvals and last changes from you.
Please ping me in GH comment when it will be ready to test (don't pay attention to `Tested-issues` label, it is only to move PR to `Contributor` column)

bitsikka commented:
> @bitsikka so waiting for approvals and last changes from you.
> Please ping me in GH comment when it will be ready to test (don't pay attention to `Tested-issues` label, it is only to move PR to `Contributor` column)

@churik understood

Update:
Turns out I needed to give equal priority to iOS and Android 😛. I was developing primarily on Android and tested only partially on iOS. So, missed some design issues(due mainly to content-layout-height calculation) including @hesterbruikman's last observations.

Fixed to a decent point now in local.

What could be better, is avoiding brief jerky transitions of the header content when keyboard or sticker panel activates. Perhaps animated transition as well later.

Todo:
- separate PR for two issues as requested by @yenda 
- explore a bit what could be better

bitsikka commented:
Update:

Fixed visual/layout problems. A lot more problems existed than the ones pointed out by @hesterbruikman. Mainly things that hadn't been considered for platform/ios? and platform/iphone-x?, but others too. Tested on those ios platforms and android as extensively as I could.

Extracted away from this PR the two fixes that are to be separate PRs as requested by @yenda. Those extra PRs(one of which is done - #7799) does not affect re-review of this PR.

bitsikka commented:
both extra PRs are done as well

https://github.com/status-im/status-react/pull/7820
fix avoid unnecessary steps when (re)starting new chat

https://github.com/status-im/status-react/pull/7802
fix chat does not scroll if touch starts on top of datemark

bitsikka commented:
> ### Latest Jenkins Builds for convenience

> ✔️	[fd86ae0](https://github.com/status-im/status-react/commit/fd86ae0057e9aa87d4b973958796d2bca9d50c44)	[#13](https://ci.status.im/job/status-react/job/prs/job/macos/job/PR-7663/13/)	2019-03-23 14:17:51	~14 min	`macos`	[📦 dmg](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190323-140337-fd86ae-pr.dmg)
> ✔️	[fd86ae0](https://github.com/status-im/status-react/commit/fd86ae0057e9aa87d4b973958796d2bca9d50c44)	[#13](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-7663/13/)	2019-03-23 14:19:34	~15 min	`ios`	[📦 ipa](https://i.diawi.com/phJaRF)
> ✔️	[fd86ae0](https://github.com/status-im/status-react/commit/fd86ae0057e9aa87d4b973958796d2bca9d50c44)	[#13](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7663/13/)	2019-03-23 14:21:31	~18 min	`android-e2e`	[📦 apk](https://status-im.ams3.digitaloceanspaces.com/StatusIm-190323-140329-fd86ae-e2e.apk)
> ✔️	[fd86ae0](https://github.com/status-im/status-react/commit/fd86ae0057e9aa87d4b973958796d2bca9d50c44)	[#13](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-7663/13/)	2019-03-23 14:21:33	~18 min	`android`	[📦 apk](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190323-140332-fd86ae-pr.apk)
> ✔️	[fd86ae0](https://github.com/status-im/status-react/commit/fd86ae0057e9aa87d4b973958796d2bca9d50c44)	[#13](https://ci.status.im/job/status-react/job/prs/job/windows/job/PR-7663/13/)	2019-03-23 14:28:45	~25 min	`windows`	[📦 exe](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190323-140339-fd86ae-pr.exe)
> ✔️	[fd86ae0](https://github.com/status-im/status-react/commit/fd86ae0057e9aa87d4b973958796d2bca9d50c44)	[#13](https://ci.status.im/job/status-react/job/prs/job/linux/job/PR-7663/13/)	2019-03-23 14:29:48	~26 min	`linux`	[📦 App](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190323-140335-fd86ae-pr.AppImage)

bitsikka commented:
ping @churik

churik commented:
Thank you @bitsikka! amazing work!
Checked on IPhone 7 (IOS 11.4.1) and LG V20 (Android 8):
- public chat with empty history
- history of public chat with some history
- empty state in 1-1 chat; update of message in case of updating name / set custom name / updating picture
- empty state of group chat when you creator or member
- recover account with one-to-one chats

Found one small issue.
**Endless "Loading" indicator is shown if public chat appeared after sync**

Steps:
- create account on device A, join any public chat
- recover this account on device B
- enable pairing between device A and device B
- sync devices
- open public chat that appeared after sync on device B

Expected result: Loading indicator is replaced with "It's been quite here..." message
Actual result: endless loading indicator. Replaced wtih text only after restarting the app.
![1232413](https://user-images.githubusercontent.com/4557972/54916781-9760ba00-4efa-11e9-9d47-42dc168eda07.png)

@hesterbruikman should we create this one separately or fix here?

bitsikka commented:
> Found one small issue.
> **Endless "Loading" indicator is shown if public chat appeared after sync**

Had not thought about this case. I think I know the reason, let me just fix it here. Had in bothering mind to add some sanity checks at couple of places anyways - which should fix it.

Basically, need to update :loading status when messages are received and :message/add(ed)

hesterbruikman commented:
> 
> 
> > Found one small issue.
> > **Endless "Loading" indicator is shown if public chat appeared after sync**
> 
> Had not thought about this case. I think I know the reason, let me just fix it here. Had in bothering mind to add some sanity checks at couple of places anyways - which should fix it.
> 
> Basically, need to update :loading status when messages are received and :message/add(ed) and a bit more stuff

@bitsikka Sounds like adding some logic, do you estimate this would cost you more than say 1 hour?

bitsikka commented:
> > > Found one small issue.
> > > **Endless "Loading" indicator is shown if public chat appeared after sync**
> > 
> > 
> > Had not thought about this case. I think I know the reason, let me just fix it here. Had in bothering mind to add some sanity checks at couple of places anyways - which should fix it.
> > Basically, need to update :loading status when messages are received and :message/add(ed) and a bit more stuff
> 
> @bitsikka Sounds like adding some logic, do you estimate this would cost you more than say 1 hour?



> @bitsikka Sounds like adding some logic, do you estimate this would cost you more than say 1 hour?

no more than an hour, but middle of wrapping up #7820 right now(also won't be long) - will get to it immediately after I wrap than one up today

bitsikka commented:
@churik *"Endless "Loading" indicator is shown if public chat appeared after sync"* case is resolved in this latest push.

I've been kind of expecting more reviews by reviewers. Feeling a bit less thorough without an explicit go-ahead from at least one core-dev.

Other than that this and accompanying PRs are ready from my side.  For overall reliability the main PR could really use the accompanying PRs. Especially #7820 without which if user re-adds an already active public chat - there would be endless loading screen.

bitsikka commented:
spoke too soon.. sorry.. endless loading is still there after re-testing on clean install right now.

investigating further.

bitsikka commented:
i see - gotta handle one more case.. working it out.

cammellos commented on src/status_im/chat/subs.cljs:
```diff
@@ -143,6 +177,17 @@
+ :<- [:chats/current-chat]
+ :<- [:chats/current-chat-messages]
+ (fn [[{:keys [might-have-join-time-messages?]} messages]]
+   (if (boolean might-have-join-time-messages?)
```
No need to coherce to bool for checking in a `if` I believe

bitsikka replied:
ok.. not going to feel funny about it anymore :)

bitsikka replied:
done

cammellos commented on src/status_im/mailserver/core.cljs:
```diff
@@ -276,9 +276,19 @@
+                                 {:keys [request-id topics]}]
   (when (:mailserver/current-request db)
-    {:db (assoc-in db [:mailserver/current-request :request-id] request-id)}))
+    (if (= 1 (count topics))
```
This is a bit fragile and it's not immediately clear what we are trying to do, probably I would factor it in a function:
`(initial-public-chat-request? ...)` so it's clearer what we are trying to do (also tests :) )

bitsikka replied:
I see why this is fragile - this is it, the piece I was missing 😁. Thanks for the hint! - had the whole mental model incomplete(still ways to go, I'm sure :p). Going to update accordingly.

bitsikka replied:
I made the naming in let binding clear instead of factoring it in a function. Hope that is acceptable.

Otherwise, this was indeed the missing piece. I had only observed that the topics in request was one in most cases before. Didn't realize that the request could have multiple topics like in the case of handling sync of public chats between devices.

cammellos commented on src/status_im/mailserver/core.cljs:
```diff
@@ -523,17 +533,37 @@
+                                        (filter #(= requestID (:join-time-mail-request-id %))))
+                                  (vals chats))
+            chat            (first new-chats)]
+        (if (seq new-chats)
```
It's probably clearer if you check for `chat` rather than `new-chats` (although it is true that if `new-chats` are empty, `chat` is also empty).
i.e
`(if chat ..)`

bitsikka replied:
got it. yes.

bitsikka replied:
done

cammellos commented on src/status_im/chat/models.cljs:
```diff
@@ -56,6 +56,14 @@
   [{:keys [current-chat-id] :as db} ui-element]
   (update-in db [:chat-ui-props current-chat-id ui-element] not))
 
+(defn join-time-messages-checked
```
I would add a description of what these fields are used for, as it might not be immediately clear

bitsikka replied:
will do

bitsikka replied:
done

cammellos reviewed (changes requested):
Looks good overall, some comments on styling etc

bitsikka commented:
going to sleep over it - mental model that was off ways bit just got updated from @cammellos review and suggestions. Going to tread carefully tomorrow and solve the tricky bits.

bitsikka commented:
@churik *'Endless "Loading" indicator is shown if public chat appeared after sync'* case is resolved for real :) in this latest push. Tested in iPhone X simulator.

plus changes requested by @cammellos

ready 🖐  as far as I know so far

cammellos commented on src/status_im/mailserver/core.cljs:
```diff
@@ -276,9 +276,28 @@
+                                 {:keys [request-id topics]}]
   (when (:mailserver/current-request db)
-    {:db (assoc-in db [:mailserver/current-request :request-id] request-id)}))
+    (let [fresh-public-chats    (into {}
```
there's a bit of duplication, you could refactor in a fn such as `chats->fresh-chats chats` or at least factor out the transducer `(comp (f....))`

bitsikka replied:
got it. will do.

bitsikka replied:
done

cammellos commented on src/status_im/mailserver/core.cljs:
```diff
@@ -523,17 +542,40 @@
                    :mailserver/pending-requests))})
 
 (fx/defn handle-request-completed
```
Could you please adding some tests specifying the behavior of this `fn` , as now there are quite a few conditional branches. Thanks!

bitsikka replied:
most certainly

bitsikka replied:
done. But I didn't know how exhaustive to make the tests. I think I've covered the essentials, but being that this my first time writing tests, I may be doing it all wrong or something :p.

bitsikka commented:
afk for a bit. Gonna finish review requests soon.

bitsikka commented:
Got all revisions completed in local except tests. Gonna work on the tests now before submitting updated PR tomorrow if not today.

bitsikka commented:
Did the changes requested by @cammellos. Ready for another review.

bitsikka commented:
Latest Jenkins Builds for convenience

> ✔️	[ae1f38e](https://github.com/status-im/status-react/commit/ae1f38e4a4c7f25008c5dd7e49ade749ae41d7cf)	[#16](https://ci.status.im/job/status-react/job/prs/job/macos/job/PR-7663/16/)	2019-03-28 18:43:05	~14 min	`macos`	[📦 dmg](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190328-182855-ae1f38-pr.dmg)
> ✔️	[ae1f38e](https://github.com/status-im/status-react/commit/ae1f38e4a4c7f25008c5dd7e49ade749ae41d7cf)	[#16](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-7663/16/)	2019-03-28 18:44:57	~16 min	`ios`	[📦 ipa](https://i.diawi.com/HxdqP4)
> ✔️	[ae1f38e](https://github.com/status-im/status-react/commit/ae1f38e4a4c7f25008c5dd7e49ade749ae41d7cf)	[#16](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-7663/16/)	2019-03-28 18:46:37	~17 min	`android`	[📦 apk](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190328-182848-ae1f38-pr.apk)
> ✔️	[ae1f38e](https://github.com/status-im/status-react/commit/ae1f38e4a4c7f25008c5dd7e49ade749ae41d7cf)	[#16](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7663/16/)	2019-03-28 18:47:58	~19 min	`android-e2e`	[📦 apk](https://status-im.ams3.digitaloceanspaces.com/StatusIm-190328-182846-ae1f38-e2e.apk)
> ✔️	[ae1f38e](https://github.com/status-im/status-react/commit/ae1f38e4a4c7f25008c5dd7e49ade749ae41d7cf)	[#16](https://ci.status.im/job/status-react/job/prs/job/windows/job/PR-7663/16/)	2019-03-28 18:54:03	~25 min	`windows`	[📦 exe](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190328-182857-ae1f38-pr.exe)
> ✔️	[ae1f38e](https://github.com/status-im/status-react/commit/ae1f38e4a4c7f25008c5dd7e49ade749ae41d7cf)	[#16](https://ci.status.im/job/status-react/job/prs/job/linux/job/PR-7663/16/)	2019-03-28 18:55:04	~26 min	`linux`	[📦 App](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190328-182853-ae1f38-pr.AppImage)

bitsikka commented:
Handled one more edge case(not unlikely for busy public chat topic) where when fresh public chat is (re)added and it is in the loading screen with loading messages from mailserver is not done.

If the chat happens to receive message(s) not requested from mailserver, but from others online active in the public chat. Loading screen ought to be gone and the message(s) shown, even though request to mailserver might not have completed by then.

I think there is nothing more to add or take away now.

ready for review.

cammellos commented on src/status_im/chat/models/message.cljs:
```diff
@@ -287,35 +287,53 @@
   (apply fx/merge cofx
          (map (partial update-last-message (:chats db)) chat-ids)))
 
+(fx/defn unfresh-public-chats
```
probably I would go for slightly different naming, rather than `unfresh`/`fresh`, I would probably change `fresh` to `never-synced` (as that is what they are correct?), fresh is a bit harder as `synced` might be a bit misleading, and `synced-at-least-once` is probably too verbose, but minor point

bitsikka replied:
yes, fresh means never-synced with mailserver or never received any messages from other client peers active in the chat after chat (re)added.

I went ahead and re-factored the naming. Also re-factored duplication and unnecessary filtering steps in chats->never-synced-public-chats.

cammellos reviewed (approved):
looks good, complicated stuff but worth the effort, well done!

bitsikka commented:
> looks good, complicated stuff but worth the effort, well done!

Thanks @cammellos!

For being my second PR, this one has been many times heftier than I would've thought. But indeed, well worth the effort over-all.

@churik @hesterbruikman and @StatusSceptre more re-reviews might just be coming in (which if it is, wouldn't hurt/would help), please coordinate in what ways to proceed.

for convenience, latest builds

> ✔️	[6fb7b93](https://github.com/status-im/status-react/commit/6fb7b939a82fab2798421c4f99d14e589e6c8beb)	[#20](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-7663/20/)	2019-03-29 15:12:20	~16 min	`ios`	[📦 ipa](https://i.diawi.com/YPwMCQ)
> ✔️	[6fb7b93](https://github.com/status-im/status-react/commit/6fb7b939a82fab2798421c4f99d14e589e6c8beb)	[#20](https://ci.status.im/job/status-react/job/prs/job/macos/job/PR-7663/20/)	2019-03-29 15:12:46	~16 min	`macos`	[📦 dmg](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190329-145615-6fb7b9-pr.dmg)
> ✔️	[6fb7b93](https://github.com/status-im/status-react/commit/6fb7b939a82fab2798421c4f99d14e589e6c8beb)	[#20](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-7663/20/)	2019-03-29 15:13:36	~17 min	`android`	[📦 apk](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190329-145601-6fb7b9-pr.apk)
> ✔️	[6fb7b93](https://github.com/status-im/status-react/commit/6fb7b939a82fab2798421c4f99d14e589e6c8beb)	[#20](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7663/20/)	2019-03-29 15:16:00	~20 min	`android-e2e`	[📦 apk](https://status-im.ams3.digitaloceanspaces.com/StatusIm-190329-145558-6fb7b9-e2e.apk)
> ✔️	[6fb7b93](https://github.com/status-im/status-react/commit/6fb7b939a82fab2798421c4f99d14e589e6c8beb)	[#20](https://ci.status.im/job/status-react/job/prs/job/linux/job/PR-7663/20/)	2019-03-29 15:16:53	~20 min	`linux`	[📦 App](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190329-145612-6fb7b9-pr.AppImage)
> ✔️	[6fb7b93](https://github.com/status-im/status-react/commit/6fb7b939a82fab2798421c4f99d14e589e6c8beb)	[#20](https://ci.status.im/job/status-react/job/prs/job/windows/job/PR-7663/20/)	2019-03-29 15:18:47	~22 min	`windows`	[📦 exe](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190329-145620-6fb7b9-pr.exe)

bitsikka commented:
found one more case for this PR

In mobile, if you click on hash-tag of public chat in any chat;

If the public chat happens to be never before added public chat - it shows blank instead of "There are no messages in this chat yet"

fix coming up

bitsikka commented:
> found one more case for this PR
> 
> In mobile, if you click on hash-tag of public chat in any chat;
> 
> If the public chat happens to be never before added public chat - it shows blank instead of "There are no messages in this chat yet"
> 
> fix coming up

✔️ done!

> ✔️	[c7d1a2b](https://github.com/status-im/status-react/commit/c7d1a2bcfdbfef08e2aa2913785e32f242c94a05)	[#21](https://ci.status.im/job/status-react/job/prs/job/macos/job/PR-7663/21/)	2019-04-01 11:28:12	~13 min	`macos`	[📦 dmg](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190401-111448-c7d1a2-pr.dmg)
> ✔️	[c7d1a2b](https://github.com/status-im/status-react/commit/c7d1a2bcfdbfef08e2aa2913785e32f242c94a05)	[#21](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-7663/21/)	2019-04-01 11:30:00	~15 min	`ios`	[📦 ipa](https://i.diawi.com/9B5kx3)
> ✔️	[c7d1a2b](https://github.com/status-im/status-react/commit/c7d1a2bcfdbfef08e2aa2913785e32f242c94a05)	[#21](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-7663/21/)	2019-04-01 11:32:24	~17 min	`android`	[📦 apk](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190401-111440-c7d1a2-pr.apk)
> ✔️	[c7d1a2b](https://github.com/status-im/status-react/commit/c7d1a2bcfdbfef08e2aa2913785e32f242c94a05)	[#21](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7663/21/)	2019-04-01 11:33:51	~19 min	`android-e2e`	[📦 apk](https://status-im.ams3.digitaloceanspaces.com/StatusIm-190401-111437-c7d1a2-e2e.apk)
> ✔️	[c7d1a2b](https://github.com/status-im/status-react/commit/c7d1a2bcfdbfef08e2aa2913785e32f242c94a05)	[#21](https://ci.status.im/job/status-react/job/prs/job/linux/job/PR-7663/21/)	2019-04-01 11:40:14	~25 min	`linux`	[📦 App](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190401-111445-c7d1a2-pr.AppImage)
> ✔️	[c7d1a2b](https://github.com/status-im/status-react/commit/c7d1a2bcfdbfef08e2aa2913785e32f242c94a05)	[#21](https://ci.status.im/job/status-react/job/prs/job/windows/job/PR-7663/21/)	2019-04-01 11:41:15	~26 min	`windows`	[📦 exe](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190401-111450-c7d1a2-pr.exe)

churik commented:
Besides of mentioned in https://github.com/status-im/status-react/pull/7663#issuecomment-476159232, on last build rechecked:
- state when chats are appearing as a result of chat sync;
- deep links to user profiles, public chats;
- links to public chats inside app.

Amazing job, @bitsikka !

bitsikka commented:
> Amazing job, @bitsikka !

Thanks! :)

churik commented:
@bitsikka would appreciate if you can take a look into https://github.com/status-im/status-react/issues/7915

bitsikka commented:
> @bitsikka would appreciate if you can take a look into #7915

on it
